### ISSUE TITLE
Enum with data is not supported properly

### ISSUE THREAD

**Issue by SkullTech** [metaplex-foundation/solita#53]
I have the following enum in my Rust code, which I'm using in an account as well as an instruction argument

```rust
#[derive(AnchorSerialize, AnchorDeserialize, Clone, PartialEq, Eq)]
pub enum CollectionInfo {
    V1 {
        symbol: String,
        verified_creators: Vec<Pubkey>,
        whitelist_root: [u8; 32],
    },
    V2 {
        collection_mint: Pubkey,
    },
}
```

Solita generates the TS type from this as

```ts
import * as beet from '@metaplex-foundation/beet'
/**
 * @category enums
 * @category generated
 */
export enum CollectionInfo {
  V1,
  V2,
}

/**
 * @category userTypes
 * @category generated
 */
export const collectionInfoBeet = beet.fixedScalarEnum(
  CollectionInfo
) as beet.FixedSizeBeet<CollectionInfo, CollectionInfo>
```

This doesn't have any of the data fields, making it pretty useless. Without this the SDK for my program becomes unusable. Is this supported by Solita? If so, what am I missing here? If not, can you add support for this?

**Comment by SkullTech** [comment]
@thlorenz 👀

**Comment by thlorenz** [comment]
You're correct, enum with inline data is not yet supported since there wasn't a use case yet.
Try to separate the inline data into structs, i.e. `V1Args` and `V2Args`.
Afaik that should work.

**Comment by thlorenz** [comment]
However the above may only work for Instruction enums.

Could you please provide me with the IDL produced by anchor so I can tackle this at some point?

**Comment by SkullTech** [comment]
This is a stripped down version of the IDL relevant to this issue. @thlorenz And yes please support this, this is supported by Anchor so many other programs might be using this too.

```json
{
  "version": "0.1.0",
  "name": "solvent",
  "instructions": [
    {
      "name": "createBucket",
      "accounts": [
        {
          "name": "signer",
          "isMut": true,
          "isSigner": true
        },
        {
          "name": "solventAuthority",
          "isMut": false,
          "isSigner": false,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "type": "string",
                "value": "authority-seed"
              }
            ]
          }
        },
        {
          "name": "bucketState",
          "isMut": true,
          "isSigner": false,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "type": "publicKey",
                "account": "Mint",
                "path": "droplet_mint"
              },
              {
                "kind": "const",
                "type": "string",
                "value": "bucket-seed-v3"
              }
            ]
          }
        },
        {
          "name": "dropletMint",
          "isMut": true,
          "isSigner": true
        },
        {
          "name": "tokenProgram",
          "isMut": false,
          "isSigner": false
        },
        {
          "name": "systemProgram",
          "isMut": false,
          "isSigner": false
        },
        {
          "name": "rent",
          "isMut": false,
          "isSigner": false
        }
      ],
      "args": [
        {
          "name": "collectionInfo",
          "type": {
            "defined": "CollectionInfo"
          }
        }
      ]
    }
  ],
  "accounts": [
    {
      "name": "BucketStateV3",
      "type": {
        "kind": "struct",
        "fields": [
          {
            "name": "bump",
            "type": "u8"
          },
          {
            "name": "dropletMint",
            "type": "publicKey"
          },
          {
            "name": "collectionInfo",
            "type": {
              "defined": "CollectionInfo"
            }
          }
        ]
      }
    }
  ],
  "types": [
    {
      "name": "CollectionInfo",
      "type": {
        "kind": "enum",
        "variants": [
          {
            "name": "V1",
            "fields": [
              {
                "name": "symbol",
                "type": "string"
              },
              {
                "name": "verified_creators",
                "type": {
                  "vec": "publicKey"
                }
              },
              {
                "name": "whitelist_root",
                "type": {
                  "array": [
                    "u8",
                    32
                  ]
                }
              }
            ]
          },
          {
            "name": "V2",
            "fields": [
              {
                "name": "collection_mint",
                "type": "publicKey"
              }
            ]
          }
        ]
      }
    }
  ],
}
```

**Comment by thlorenz** [comment]
Just letting you know I started work on this, but it turns out a bit more complex than I thought.

The TypeScript type would probably be a discriminated union since there are no data enums in JS or TypeScript:

```ts
import * as web3 from '@solana/web3.js'

export type CollectionInfoV1 = {
  __kind: 'V1'
  symbol: string
  verifiedCreators: web3.PublicKey
  whiteListRoot: number[] /* 32 */
}

export type CollectionInfoV2 = {
  __kind: 'V2'
  collectionMint: web3.PublicKey
}

export type CollectionInfo = CollectionInfoV1 | CollectionInfoV2
```

and would have guards going along with it:

```ts
export const isCollectionInfoV1 = (x: CollectionInfo): x is CollectionInfoV1 =>
  x.__kind === 'V1'
export const isCollectionInfoV2 = (x: CollectionInfo): x is CollectionInfoV2 =>
  x.__kind === 'V2'
```
This also allows type-safe variant matching:

```ts
function handleCollectionInfo(info: CollectionInfo) {
  switch (info.__kind) {
    case 'V1':
      // TS narrowws to CollectionInfoV1
      console.log(info.symbol)
      break
    case 'V2':
      // TS narrowws to CollectionInfoV2
      console.log(info.collectionMint)
      break
  }
}
```
Combined this is as close as one probably gets to how those work in Rust.

* * *

However the de/serializer is a bit more complex since it basically has to branch depending on the variant.

I'll have to add this to beet first.

In the meantime please double check that the above type would be what you'd want.

**Comment by thlorenz** [comment]
Reposting from Linear as the issue comments don't sync:

* * *
At this point solita generates the below code for your specific use case:

```ts
import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
/**
 * This type is used to derive the {@link CollectionInfo} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link CollectionInfo} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type CollectionInfoRecord = {
  V1: {
    symbol: string
    verifiedCreators: web3.PublicKey[]
    whitelistRoot: number[] /* size: 32 */
  }
  V2: { collectionMint: web3.PublicKey }
}

/**
 * Union type respresenting the CollectionInfo data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isCollectionInfo*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type CollectionInfo = beet.DataEnumKeyAsKind<CollectionInfoRecord>

export const isCollectionInfoV1 = (
  x: CollectionInfo
): x is CollectionInfo & { __kind: 'V1' } => x.__kind === 'V1'
export const isCollectionInfoV2 = (
  x: CollectionInfo
): x is CollectionInfo & { __kind: 'V2' } => x.__kind === 'V2'

/**
 * @category userTypes
 * @category generated
 */
export const collectionInfoBeet = beet.dataEnum<CollectionInfoRecord>([
  [
    'V1',
    new beet.FixableBeetArgsStruct<CollectionInfoRecord['V1']>(
      [
        ['symbol', beet.utf8String],
        ['verifiedCreators', beet.array(beetSolana.publicKey)],
        ['whitelistRoot', beet.uniformFixedSizeArray(beet.u8, 32)],
      ],
      'CollectionInfoRecord["V1"]'
    ),
  ],

  [
    'V2',
    new beet.BeetArgsStruct<CollectionInfoRecord['V2']>(
      [['collectionMint', beetSolana.publicKey]],
      'CollectionInfoRecord["V2"]'
    ),
  ],
]) as beet.FixableBeet<CollectionInfo>
```

About to publish the beet and solita updates that include this feature but posting here so you can have a look in the meantime.

**Comment by thlorenz** [comment]
Published as v0.9.0. @SkullTech please verify that your issue is fixed and if not and/or if you have improvement suggestions we'll re-open this issue.
